com.unity.robotics.ros-tcp-connector/Tests/Runtime/TFSystemTests.cs

```
@@ -0,0 +1,185 @@
+using System.Collections;
+using System.Collections.Generic;
+using NUnit.Framework;
+using RosMessageTypes.BuiltinInterfaces;
+using RosMessageTypes.Geometry;
+using RosMessageTypes.Tf2;
+using RosMessageTypes.Std;
+using Unity.Robotics.ROSTCPConnector;
+using Unity.Robotics.ROSTCPConnector.MessageGeneration;
+using Unity.Robotics.ROSTCPConnector.ROSGeometry;
+using UnityEngine;
+using UnityEngine.TestTools;
+
+namespace Tests.Runtime
+{
+    public class TFSystemTests
+    {
+        const string simple_frame_id = "test_frame_id";
+        const string simple_frame_id2 = "test_frame_id2";
+        const string parent_frame_id = "parent_frame_id";
+        const string composite_frame_id = parent_frame_id + "/" + simple_frame_id;
+        const string composite_frame_id2 = parent_frame_id + "/" + simple_frame_id2;
+
+        HeaderMsg MakeHeaderMsg(TimeMsg time, string frameID)
+        {
+#if !ROS2
+            return new HeaderMsg((uint)0, time, frameID);
+#else
+            return new HeaderMsg(time, frameID);
+#endif
+        }
+
+        TimeMsg MakeTimeMsg(int secs, int nsecs)
+        {
+#if !ROS2
+            return new TimeMsg((uint)secs, (uint)nsecs);
+#else
+            return new TimeMsg(secs, (uint)nsecs);
+#endif
```

Same deal here! No action req. for the PR but I could add a backlog ticket for it

---

Hmm, for the headers, it seems reasonable to just ignore the extra/missing parameter. But for TimeMsg (and DurationMsg) we'd be casting between uint and int which seems like a dangerous thing to do silently.

---

Agreed - that could be called out as well in the "Hey you used the constructor for the other ROS warning" -- 99% of the time I think it'd be a safe cast since time is not allowed to be negative in either version.

---

Hmm, judging by this thread a Duration is allowed to be negative, but a negative Time is illegal, because they want 0 to mean uninitialized.
https://github.com/ros2/rclcpp/issues/525

---

Right and since they are different classes I think we're safe to cast one to the other - if we really wanted to we could throw an exception in the case of a negative int.  I'll put a ticket in the backlog for it